### PR TITLE
Use DiplomatSlice/etc instead of splitting params into data/len pairs

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes https://github.com/rust-diplomat/diplomat/issues/613

Things this does not do that it should, but I might do in a followup:

 - [ ] Potentially get rid of DiplomatStr in favor of the new types. It's not really worth it to allow people to specify `&[&DiplomatStr]`, they should just go all the way IMO.
 - [x] Add checks so that the shortcuts are only allowed in params, not struct fields
 - [ ] Clean up the tests: https://github.com/rust-diplomat/diplomat/issues/617
 - [x] Use a StdlibOrRuntime enum instead of a boolean

### CODE REVIEW

File: runtime/src/slices.rs [hunk 1]
```diff
@@ -0,0 +1,236 @@
+use alloc::boxed::Box;
+use core::marker::PhantomData;
+use core::ops::{Deref, DerefMut};
+
+/// This is equivalent to &[T], except it has a stable repr(C) layout
+#[repr(C)]
+pub struct DiplomatSlice<'a, T> {
+    // Invariant: ptr is a valid ptr to the beginning of an &[T] allocation
```
Comment (robertbastian) [inline runtime/src/slices.rs:8]: I think this needs work around empty slices. Rust doesn't require an allocation (even an empty one) for empty slices, and Diplomat currently allows passing `NULL, 0`.

File: runtime/src/slices.rs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        }
+    }
+}
+impl<'a, T> Copy for DiplomatSlice<'a, T> {}
+
+impl<'a, T> From<&'a [T]> for DiplomatSlice<'a, T> {
+    fn from(x: &'a [T]) -> Self {
+        // Safe to construct since we're constructing it from a slice
+        DiplomatSlice {
+            ptr: x as *const [T] as *const T,
```
Comment (robertbastian) [inline runtime/src/slices.rs:30]:
```suggestion
            ptr: x.as_ptr(),
```

File: runtime/src/slices.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            core::slice::from_raw_parts_mut(self.ptr, self.len)
+        }
+    }
+}
+
+#[repr(C)]
+pub struct DiplomatOwnedSlice<T> {
+    // Invariant: ptr is a valid ptr to the beginning of an owned Box<[T]> allocation
+    ptr: *mut T,
+    // Invariant: the allocation contains at least `len` elements
```
Comment (robertbastian) [inline runtime/src/slices.rs:108]: this should be exact, because we will free it

File: runtime/src/slices.rs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        // Safety: The type invariants allow us to treat is as a valid Box<[T]>
+        unsafe { core::slice::from_raw_parts_mut(self.ptr, self.len) }
+    }
+}
+
+/// This is equivalent to &str, except it has a stable repr(C) layout
+// Safety invariant: contained slice must be valid UTF-8
+#[repr(transparent)]
+#[derive(Copy, Clone)]
+pub struct DiplomatUTF8StrSlice<'a>(DiplomatSlice<'a, u8>);
```
Comment (robertbastian) [inline runtime/src/slices.rs:164]:
```suggestion
pub struct DiplomatUtf8StrSlice<'a>(DiplomatSlice<'a, u8>);
```

File: feature_tests/dart/lib/src/BorrowedFields.g.dart [hunk 5]
```diff
@@ -43,11 +43,17 @@ final class BorrowedFields {
   static BorrowedFields fromBarAndStrings(Bar bar, String dstr16, String utf8Str) {
     final dstr16Arena = _FinalizedArena();
     final (dstr16Data, dstr16Length) = dstr16._utf16AllocIn(dstr16Arena.arena);
+    final dstr16 = ffi.Struct.create<_SliceUtf16>();
+    dstr16._data = dstr16Data;
+    dstr16._length = dstr16Length;
     final utf8StrArena = _FinalizedArena();
     final (utf8StrData, utf8StrLength) = utf8Str._utf8AllocIn(utf8StrArena.arena);
+    final utf8Str = ffi.Struct.create<_SliceUtf8>();
```
Comment (robertbastian) [inline feature_tests/dart/lib/src/BorrowedFields.g.dart:51]: I'd push the struct creation into `allocIn`, but I can follow up

File: feature_tests/dart/lib/src/BorrowedFields.g.dart [hunk 6]
```diff
@@ -75,6 +81,6 @@ final class BorrowedFields {
 }
 
 @meta.ResourceIdentifier('BorrowedFields_from_bar_and_strings')
-@ffi.Native<_BorrowedFieldsFfi Function(ffi.Pointer<ffi.Opaque>, ffi.Pointer<ffi.Uint16>, ffi.Size, ffi.Pointer<ffi.Uint8>, ffi.Size)>(isLeaf: true, symbol: 'BorrowedFields_from_bar_and_strings')
+@ffi.Native<_BorrowedFieldsFfi Function(ffi.Pointer<ffi.Opaque>)>(isLeaf: true, symbol: 'BorrowedFields_from_bar_and_strings')
```
Comment (robertbastian) [inline feature_tests/dart/lib/src/BorrowedFields.g.dart:84]: issue: the arguments disappeared

File: feature_tests/dart/lib/src/BorrowedFields.g.dart [hunk 7]
```diff
@@ -75,6 +81,6 @@ final class BorrowedFields {
...
 // ignore: non_constant_identifier_names
-external _BorrowedFieldsFfi _BorrowedFields_from_bar_and_strings(ffi.Pointer<ffi.Opaque> bar, ffi.Pointer<ffi.Uint16> dstr16Data, int dstr16Length, ffi.Pointer<ffi.Uint8> utf8StrData, int utf8StrLength);
+external _BorrowedFieldsFfi _BorrowedFields_from_bar_and_strings(ffi.Pointer<ffi.Opaque> bar, ffi.Uint16 dstr16, ffi.Uint8 utf8Str);
```
Comment (robertbastian) [inline feature_tests/dart/lib/src/BorrowedFields.g.dart:86]: issue: the types are wrong

**Review by robertbastian — COMMENTED** [review]
not an in-depth review

**Comment by Manishearth** [comment]
I'm signing off soon, but @robertbastian please do feel free to polish up the Dart section and push, I did a quick-and-dirty fix which still appears to be broken.

(let me know if you've pushed though, so I don't overwrite)

File: feature_tests/dart/lib/src/BorrowedFields.g.dart (hunk 6 above)
Comment (Manishearth) [reply feature_tests/dart/lib/src/BorrowedFields.g.dart]: Fixed. The `param_types_ffi_cast` etc stuff is kinda confusing, we generate a lot of different things in one function. I didn't bother to try and understand it too deeply but we should document this better if we get the chance.

(Also now that this is no longer doing two-parameter stuff we probably can clean things up)

File: feature_tests/dart/lib/src/BorrowedFields.g.dart (hunk 5 above)
Comment (Manishearth) [reply feature_tests/dart/lib/src/BorrowedFields.g.dart:51]: I'd love that yeah.

**Comment by Manishearth** [comment]
I tried to fix some things but Dart's still broken

**Comment by ambiguousname** [comment]
I'll take a look and adjust the JS backend. We have a few open issues with how JS handles slices anyways (I believe to do with cleanup in particular), so maybe I can re-write the JS slices system to account for those issues in addition to your changes. I'll ping you by EOD today when I've got something.

**Comment by ambiguousname** [comment]
@Manishearth I did some review of the updated `.wasm` output and how JS handles slices; I believe JS handles slices the same whether they're being used as struct fields or function parameters.

WASM output still requires a pointer and size, the new types are ignored:  `(func $icu4x_Locale_new_mv1 (type 3) (param i32 i32) (result i32)`

As for backend logic, JS uses the same `DiplomatBuf` type for slices regardless of use in . 

For clarity, struct field conversion is done in `js/type_generation/mod.rs`, line 165 and parameter conversion is done `js/type_generation/mod.rs`, line 307. Both use the same `gen_js_to_c_for_type` function.

The only change I can think of is renaming `DiplomatBuf` to `DiplomatSliceBuf` for clarity?

As a caveat however, I can't seem to find tests in `feature_tests` that handle cases where a struct has a `DiplomatSlice` as a member, so I can't verify JS behavior.

TL;DR JS already handles slices for structs and function parameters in the same way. If I'm missing something or you'd like additional changes, let me know.

**Comment by Manishearth** [comment]
Oh yeah, that makes sense, the WASM abi _hates_ structs so everything is done splatted out anyway. Nice to not have to worry about this!

**Comment by Manishearth** [comment]
And CI passes!

**Comment by Manishearth** [comment]
That took less time than I expected.

As a bonus, this simplifies the macro code so a lot of the gnarly DiplomatOption handling I wanted to do becomes really straightforward, and that reduces the amount of macro mess that was going to be in https://github.com/rust-diplomat/diplomat/pull/558 ! (I never pushed the macro mess code because I was annoyed with it)

File: core/src/ast/types.rs [hunk 8]
```diff
@@ -319,6 +319,16 @@ impl Mutability {
     }
 }
 
+/// For types like `Result`/`DiplomatResult`, `&[T]`/`DiplomatSlice<T>` which can be
+/// specified using (non-ffi-safe) Rust stdlib types, or FFI-safe `repr(C)` types from
+/// `diplomat_runtime`, this tracks which of the two were used.
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize, Debug)]
+#[allow(clippy::exhaustive_enums)] // This can only have two values
+pub enum StdlibOrDiplomat {
```
Comment (robertbastian) [inline core/src/ast/types.rs:327]: below you have `get_diplomat_slice_type` and `get_rust_slice_type` methods, but here it's `StdLib` and `Diplomat`. align?

File: example/c/main.c [hunk 9]
```diff
@@ -33,7 +33,9 @@ int main(int argc, char *argv[]) {
 
     print_decimal(fd);
 
-    Locale* locale = icu4x_Locale_new_mv1("bn", 2);
+    DiplomatStringView bn = {"bn", 2};
+
+    Locale* locale = icu4x_Locale_new_mv1(bn);
```
Comment (robertbastian) [inline example/c/main.c:38]: nit: I think in examples we should inline slices, and maybe use named fields

**Review by robertbastian — APPROVED** [review]
LG overall. Will work on the Dart simplifications so please no more force pushes.

**Comment by Manishearth** [comment]
Sounds good. I'll try to merge as soon as possible to avoid you needing to patch stack

File: example/c/main.c [hunk 10]
```diff
@@ -33,9 +33,7 @@ int main(int argc, char *argv[]) {
 
     print_decimal(fd);
 
-    DiplomatStringView bn = {"bn", 2};
-
-    Locale* locale = icu4x_Locale_new_mv1(bn);
+    Locale* locale = icu4x_Locale_new_mv1((DiplomatStringView){.data = "bn", .len = 2});
```
Comment (robertbastian) [inline example/c/main.c:36]: does this really need the cast? 😕

Comment (Manishearth) [reply example/c/main.c:36]: it fails to _parse_ without the cast. I don't understand it either.

C is weird and has terrible errors, if you can figure out how to avoid the cast be my guest 🙃🙃

Comment (robertbastian) [reply example/c/main.c:36]: ew. in that case I'm not sure if the inline is actually nicer 🤷

Comment (Manishearth) [reply example/c/main.c:36]: yeah feel free to change it back if you'd like. I'm very ambivalent here
